yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnTimelineUtils.scala

```diff
@@ -0,0 +1,720 @@
+    sourceObj.obj.foreach(f => map.put(f._1, toJavaObject(f._2)))
+    map
+  }
+
+  /**
+   * Convert a timeline event to a spark one. Includes
+   * some basic checks for validity of event payload.
+   * @param event timeline event
+   * @return an unmarshalled event
+   */
+  def toSparkEvent(event: TimelineEvent): SparkListenerEvent = {
+    val info = event.getEventInfo()
+    if (info == null) {
+      throw new IOException(E_NO_EVENTINFO)
+    }
+    if (info.size() == 0) {
+      throw new IOException(E_EMPTY_EVENTINFO)
+    }
+    val payload = toJValue(info)
+    def jsonToString: String = {
+      val json = compact(render(payload))
+      val limit = 256
+      if (json.length < limit) {
+        json
+      } else {
+          json.substring(0, limit) + " ... }"
+        }
+    }
+    logDebug(s"payload is ${jsonToString}")
+    val eventField = payload \ "Event"
+    if (eventField == JNothing) {
+      throw new IOException(s"No 'Event' entry in $jsonToString")
+    }
+
+    // now the real unmarshalling
+    try {
+      JsonProtocol.sparkEventFromJson(payload)
+    } catch {
+      // failure in the marshalling; include payload in the message
+      case ex: MappingException => {
+        logError(s"$ex while rendering $jsonToString", ex)
+        throw ex
+      }
+    }
+  }
+
+  /**
+   * Convert a spark event to a timeline event
+   * @param event handled spark event
+   * @return a timeline event
+   */
+  def toTimelineEvent(event: HandleSparkEvent): TimelineEvent = {
+    val tlEvent = new TimelineEvent()
+    tlEvent.setEventType(Utils.getFormattedClassName(event.sparkEvent)
+        + "-" + YarnTimelineUtils.uid.incrementAndGet.toString)
+    tlEvent.setTimestamp(event.time)
+    val kvMap = new JHashMap[String, Object]()
+    val json = JsonProtocol.sparkEventToJson(event.sparkEvent)
+    val jObject = json.asInstanceOf[JObject]
+    // the timeline event wants a map of java objects for Jackson to serialize
+    val hashMap = toJavaMap(jObject)
+    tlEvent.setEventInfo(hashMap)
+    tlEvent
+  }
+
+  /**
+   * Describe the event for logging
+   * @param event timeline event
+   * @return a description
+   */
+  def describeEvent(event: TimelineEvent): String = {
+    var sparkEventDetails = ""
+    try {
+      sparkEventDetails = toSparkEvent(event).toString
+    } catch {
+      case _: Exception =>
+        sparkEventDetails = "(cannot convert event details to spark exception)"
+    }
+    s"${event.getEventType()} @ ${new Date(event.getTimestamp())}" +
+        s"\n    ${sparkEventDetails}"
+  }
+
+  /**
+   * Create details of a timeline entity, by describing every event inside it
+   * @param entity entity containing a possibly empty or null list of events
+   * @return a list of event details, with a newline between each one
+   */
+  def eventDetails(entity: TimelineEntity): String = {
+    val events: util.List[TimelineEvent] = entity.getEvents
+    if (events != null) {
```

Comment:
I was getting some IDE warnings about the automatic conversions being deprecated. If they're ok in the source then I'll use them there & in a couple of other places.